kernel/src/mm/guestmem.rs

```
@@ -271,6 +278,86 @@ impl<T: Copy> InsnMachineMem for GuestPtr<T> {
     }
 }
 
+#[derive(Debug)]
+pub struct UserPtr<T: Copy> {
+    guest_ptr: GuestPtr<T>,
+}
+
+impl<T: Copy> UserPtr<T> {
+    #[inline]
+    pub fn new(v: VirtAddr) -> Self {
+        Self {
+            guest_ptr: GuestPtr::new(v),
+        }
+    }
+
+    fn check_bounds(&self) -> bool {
+        let v = VirtAddr::from(self.guest_ptr.ptr);
+
+        (USER_MEM_START..USER_MEM_END).contains(&v)
+            && (USER_MEM_START..USER_MEM_END).contains(&(v + size_of::<T>()))
+    }
+
+    #[inline]
+    pub fn read(&self) -> Result<T, SvsmError>
+    where
+        T: FromBytes,
+    {
+        if !self.check_bounds() {
+            return Err(SvsmError::InvalidAddress);
+        }
+        unsafe { self.guest_ptr.read() }
+    }
+
+    #[inline]
+    pub fn write(&self, buf: T) -> Result<(), SvsmError> {
+        if !self.check_bounds() {
+            return Err(SvsmError::InvalidAddress);
+        }
+        self.write_ref(&buf)
+    }
+
+    #[inline]
+    pub fn write_ref(&self, buf: &T) -> Result<(), SvsmError> {
+        if !self.check_bounds() {
+            return Err(SvsmError::InvalidAddress);
+        }
+        unsafe { self.guest_ptr.write_ref(buf) }
+    }
+
+    #[inline]
+    pub const fn cast<N: Copy>(&self) -> UserPtr<N> {
+        UserPtr {
+            guest_ptr: self.guest_ptr.cast(),
+        }
+    }
+
+    #[inline]
+    pub fn offset(&self, count: isize) -> UserPtr<T> {
+        UserPtr {
+            guest_ptr: self.guest_ptr.offset(count),
+        }
+    }
+}
+
+impl UserPtr<c_char> {
+    /// Reads a null-terminated C string from the user space.
+    /// Allocates memory for the string and returns a `String`.
+    pub fn read_c_string(&self) -> Result<String, SvsmError> {
+        let mut buffer = Vec::with_capacity(4096);
+
+        for offset in 0..buffer.capacity() {
```

We should return an error if the string exceeds the capacity. Also, we probably shouldn't default to using 4k as the capacity for the vector. Most strings are probably way smaller than that, the default allocation strategy of vectors should work fine.

---

I see your point, how about this? Also, introduced a new const for max path len.
```
@@ -344,17 +342,17 @@ impl UserPtr<c_char> {
     /// Reads a null-terminated C string from the user space.
     /// Allocates memory for the string and returns a `String`.
     pub fn read_c_string(&self) -> Result<String, SvsmError> {
-        let mut buffer = Vec::with_capacity(4096);
+        let mut buffer = Vec::new();
 
-        for offset in 0..buffer.capacity() {
+        for offset in 0..PATH_MAX {
             let current_ptr = self.offset(offset as isize);
             let char_result = current_ptr.read()?;
             match char_result {
-                0 => break,
+                0 => return String::from_utf8(buffer).map_err(|_| SvsmError::InvalidUtf8),
                 c => buffer.push(c as u8),
             }
         }
-        String::from_utf8(buffer).map_err(|_| SvsmError::InvalidUtf8)
+        Err(SvsmError::InvalidBytes)
     }
 }
```

---

Looks good to me.

---

Thanks, done.